### PR TITLE
Mamba Elite Extended Support

### PR DESCRIPTION
**Pull request by beranat**
The [Razer Mamba Elite mouse](https://www.razer.com/gaming-mice/razer-mamba-elite) is supported in driver and daemon.

Use [customized polychromatic](https://github.com/beranat/polychromatic) for user-friendly colors control

Difference with "IvanVojtko's patch" is effects support.

### CODE REVIEW

**Comment by nightsky30** [comment]
@beranat I like the left/right wrapper functions :+1:

**Comment by beranat** [comment]
Dear *nightsky30*, you may customise its colors as you wish in polychromatic:
image

**Comment by samindaperamuna** [comment]
Can we get this pulled into the original repository please.

File: Makefile
Comment (z3ntu) [inline Makefile]: Any explanation why?

File: daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py
Comment (z3ntu) [inline daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py]: These methods seem to be exactly the same as in `dbus_services/dbus_methods/nagahexv2.py` , why are they duplicated? (if you got confused by the nagahexv2 name, just ignore that. The naming conventions in this project are very legacy-driven :wink: )

File: daemon/openrazer_daemon/hardware/mouse.py
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: Please put multiple strings into a single line, that way we don't have a giant block of function names here (but you can leave the comments of course :) )

File: daemon/openrazer_daemon/hardware/mouse.py
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py:1381]: good idea, I wonder why I've never done that for other devices^^

File: driver/razerchromacommon.h
Comment (z3ntu) [inline driver/razerchromacommon.h]: Please don't reformat existing files without any code changes, so please revert those changes in this file and in several other files.

File: pylib/openrazer/_fake_driver/razermambaelite.cfg
Comment (z3ntu) [inline pylib/openrazer/_fake_driver/razermambaelite.cfg]: Indentation is wrong here and on line 6

File: update.sh
Comment (z3ntu) [inline update.sh]: Remove the script

**Review by z3ntu — CHANGES_REQUESTED** [review]
See comments, please remove all unnecessary code style changes.

File: Makefile
Comment (beranat) [reply Makefile]: The 'kernel/drivers/hid' is used as part of kernel's source tree, but the openrazer driver is not in, so It better to move to 'extra'. 

But when the driver's code will be merged into kernel - the path should be fixed.

File: daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py
Comment (beranat) [reply daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py]: OK, so I will fix

File: daemon/openrazer_daemon/hardware/mouse.py
Comment (beranat) [reply daemon/openrazer_daemon/hardware/mouse.py]: OK

File: daemon/openrazer_daemon/hardware/mouse.py
Comment (beranat) [reply daemon/openrazer_daemon/hardware/mouse.py:1381]: Wellcome 8-)

File: driver/razerchromacommon.h
Comment (beranat) [reply driver/razerchromacommon.h]: OK, But according [Linux kernel coding style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html) and practice - the line's width 80 is a really better.

File: update.sh
Comment (beranat) [reply update.sh]: OK, this file sneaks into repository

**Comment by samindaperamuna** [comment]
Good job!

File: Makefile
Comment (z3ntu) [reply Makefile]: I'd rather not change it as it's working nicely for multiple years :)

File: daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py
Comment (z3ntu) [inline daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py:24]: Do we need those "fake" methods? We have scroll, logo, left side & right side zones, so why add a zone being both left&right?

**Review by z3ntu — CHANGES_REQUESTED** [review]
There are still a bunch of whitespace/autoformat changes in several `.c` files, please revert those.

File: Makefile
Comment (beranat) [reply Makefile]: Let's replace our pentium i-core to i80286 - it works for multiple years :-((
I see no reason to argue - my target is "push merge to master brunch for all mamba elite's users" and this is not the last legacy in the project, so let it be.

File: daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py
Comment (beranat) [reply daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py:24]: Method "get_brightness" complements "set_brightness" (set brightness level for all components), Another devices has both methods, so implementing this one does simplifiy control application.

Methods "*_effect" using for simplify and unify code

Comment (z3ntu) [reply daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py:24]: In my opinion the driver/daemon shouldn't make opaque wrappers that frontends have no idea what they are doing. If you want to "synchronize" all lighting zones on the device in the frontend, it should be done in the frontend and not via some wrapper behind that.

Comment (beranat) [reply daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py:24]: Absolutely agree! You must not smoking on gas station - and the code has not been placed in kernel module. 

It should located in high level wrapper, which are used for simplify and unify access and client (like X11 vs Qt) may choose the best way between:
* Direct access to hardware (/sys/...) - low level (fast/short but needs peripheral's knowledge)
* D-bus access - Unified _High level API_ (offer simple programming interface and decreases device-specific stuff).

Comment (z3ntu) [reply daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py:24]: That's not what I have meant and I'm pretty sure you know that...

In short: I'm the maintainer, if you want to get the pull request merged please accept my position as maintainer and implement the requested changes otherwise I won't merge it.

Comment (Smarre) [reply daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py:24]: Or probably we just need to do a fork instead of all this silly nitpicking —.—

**Comment by samindaperamuna** [comment]
Excellent progress so far. Any update on the review?
Thanks

File: daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py
Comment (beranat) [reply daemon/openrazer_daemon/dbus_services/dbus_methods/mambaelite.py:24]: Dear **@Smarre**, There is no problem to fork (and I have already fork the repository for  Mamba Elite development) - but maintainers support another devices so prefer to merge into master brunch, and for this task I ready to pull down code development requirements in *optional cases* (like [Linux kernel coding style](https://01.org/linuxgraphics/gfx-docs/drm/process/coding-style.html) - the specifications has been wrote by more experience and smart people) if maintainer wants - it is not good, but acceptable and users would get solution.
 
Dear **The Maintainer @z3ntu**, I would thank you for working and coordinate the openrazer project. But as you notice  (and for my luck) - You are *just* maintainer and I may set aside to do controversial solutions - the library's (intermediate layer's) target is to simplify and unify.

**Comment by beranat** [comment]
Dear **@samindaperamuna**, not yet but may be later when polychromatic-1.0.0 has been released

**Comment by z3ntu** [comment]
I've cleaned up the patch and removed the wrappers which I didn't like - but it's merged now.

**Comment by sypherus** [comment]
Hello, why can't i control the right and left side effects?

**Comment by z3ntu** [comment]
@sypherus Probably the GUI(s) you're using don't support left/right side leds yet. 

I'm hoping to release a new version of RazerGenie with support for that in the near future. Not sure if the latest dev version of polychromatic supports those leds yet.

**Comment by o0mario0o** [comment]
Hi, all

My mamba elite is not recognized by Polychromatic, any idea what can be wrong?

I'm using daily builds for Polychromatic (v 0.3.12) and have OpenRazer (v 2.7)

I am running this on Ubuntu 20.04, any help is appreciated.

Thanks

**Comment by o0mario0o** [comment]
@beranat  , @z3ntu 

Hi, I am having the same issues listed by @sypherus .

I tried using the latest Polychromatic and RazerGenie, with OpenRazer (v 2.8).

Neither program allows for the editing of the light strips on the sides of the mouse.

Any updates or support for this?

(please correct me if it has been fixed already)

**Comment by beranat** [comment]
Dear @o0mario0o  and @sypherus , so sorry for long delay, but I will describe the possible issues:

The driver and its installation (some kernels needs a signed driver)
Could you check the driver (from root or sudo)
* ```modinfo razermnouse``` the driver is available
* ```modprobe razermnouse``` the driver can be loaded

The device is available for application (unprivileged user)
```cat /sys/bus/hid/drivers/razermouse/*/hidraw/hidraw0/device/device_type```
Should shows 'Razer Mamba Elite', some Linuxes (Fedora/CentOS) do not use 'plugdev' group for attachable devices (they prefer input)

After these fixes RazerGenie (master) + openrazer (2.8 from hardware:razer-repo) work with effects (like static/spectrum and etc could be defined for all mouse parts - logo/wheel/left/right sides), but no custom color.

**Comment by beranat** [comment]
About more deep and layouts:

The source patch has been pull requested to (Polychromatic 0.3.x)[https://github.com/beranat/polychromatic] and (OpenRazer 2.6)[https://github.com/beranat/openrazer].

Polychromatic is under global code rewrite for v1.0.0 [read this post for details](https://github.com/polychromatic/polychromatic/pull/200#issuecomment-576198716), so you could use  [old but more stable 0.3.x fork](https://github.com/beranat/polychromatic), I hope as **lah7** stabilizes version the patch will be added.

OpenRazer commits into master only kernel driver code (that likes the maintenancer) and even patched Polychromatic will not work without patching the python-library or re-implement wrappers in old Polychromatic (waste of time, it is better waiting code refactor).

So solutions are:
1. Use not last, but stable  (Polychromatic)[https://github.com/beranat/polychromatic] and (OpenRazer)[https://github.com/beranat/openrazer].
2. Modify Polychromatic or Openrazer
3. Wait for RazerGenie/Polychromatic update or 

P.S. I am using 1 + 3 now, because I need not the last versions for my hardware.

**Comment by o0mario0o** [comment]
Hi @beranat ,

Thank you for your support! 

I tried the modinfo/probe in terminal, and it ran without errors.
I also went into the file you mentioned and it says 'Razer Mamba Elite'.

I removed my user from the plugdev group but openrazer wouldn't run, so
I reverted back to adding it to plugdev.

If it helps, I am also using Ubuntu 20.04, which doesn't have python2 by default.

I think I will wait for Polychromatic 1.0.0, and fresh install everything.

Thank you for your work and contributions, nonetheless; this is a great project.

**Comment by beranat** [comment]
Dear @o0mario0o, I will check on Ubuntu 20.04 LTS and report you

**Comment by o0mario0o** [comment]
@beranat 

Any update on how things are going on the project and issue? 

I still have the same issue, so researching for workarounds once again, haha.

**Comment by o0mario0o** [comment]
> Dear _nightsky30_, you may customise its colors as you wish in polychromatic:
> image

Hi @beranat 

How can I use this UI and/or version of Polychromatic to customize left/right strips on Mamba Elite?
(doesn't have to be stable, just able to get the job done, haha).

**Comment by Giuseppe499** [comment]
> Dear @o0mario0o and @sypherus , so sorry for long delay, but I will describe the possible issues:
> 
> The driver and its installation (some kernels needs a signed driver)
> Could you check the driver (from root or sudo)
> 
>     * `modinfo razermnouse` the driver is available
> 
>     * `modprobe razermnouse` the driver can be loaded
> 
> 
> The device is available for application (unprivileged user)
> `cat /sys/bus/hid/drivers/razermouse/*/hidraw/hidraw0/device/device_type`
> Should shows 'Razer Mamba Elite', some Linuxes (Fedora/CentOS) do not use 'plugdev' group for attachable devices (they prefer input)
> 
> After these fixes RazerGenie (master) + openrazer (2.8 from hardware:razer-repo) work with effects (like static/spectrum and etc could be defined for all mouse parts - logo/wheel/left/right sides), but no custom color.

Hi @beranat ,

Thank you for your support!

I tried the modinfo/probe in terminal, and it ran without errors but when i try to open the file "/sys/bus/hid/drivers/razermouse/*/hidraw/hidraw0/device/device_type" that you mentioned it says that the file does not exist. Instead, when I open the file "/sys/bus/hid/drivers/razermouse/0003:1532:006C.0004/hidraw/hidraw2/device/device_type" it shows "Razer Mamba Elite".

When i try to use RazerGenie i can only configure Logo and Scroll Lighting. If i try to use the button "Open custom editor" it says "Device type not implemented" and turn off all mouse lighting. What am I doing wrong?

Here are the screenshots of RazerGenie
Istantanea_2020-10-21_12-34-12
Istantanea_2020-10-21_12-34-27

If it can be usefull,  I am using xUbuntu 20.04.

Thank you for your work and contributions, this is a really great project and I can finally have a cool mouse with linux drivers support.

**Pull request closed without merging**
